File: app/src/main/java/org/autojs/autojs/ui/log/LogActivity.java
package org.autojs.autojs.ui.log;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.Nullable;

import org.autojs.autojs.AutoJs;
import org.autojs.autojs.core.console.ConsoleView;
import org.autojs.autojs.core.console.GlobalConsole;
import org.autojs.autojs.ui.BaseActivity;
import org.autojs.autojs.util.ViewUtils;
import org.autojs.autojs6.R;
import org.autojs.autojs6.databinding.ActivityLogBinding;
import org.autojs.autojs6.databinding.ConsoleViewBinding;
import org.jetbrains.annotations.NotNull;

public class LogActivity extends BaseActivity {

    private GlobalConsole mConsoleImpl;

    private final int RESULT_CODE_EXPORT = 33128;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ActivityLogBinding activityLogBinding = ActivityLogBinding.inflate(getLayoutInflater());
        setContentView(activityLogBinding.getRoot());

        setToolbarAsBack(R.string.text_log);

        ConsoleView mConsoleView = activityLogBinding.console;

        mConsoleImpl = AutoJs.getInstance().getGlobalConsole();
        mConsoleView.setConsole(mConsoleImpl);
        mConsoleView.setLogActivity(this);
        mConsoleView.setPinchToZoomEnabled(true);

        ConsoleViewBinding consoleViewBinding = ConsoleViewBinding.inflate(getLayoutInflater());
        consoleViewBinding.inputContainer.setVisibility(View.GONE);

        activityLogBinding.fab.setOnClickListener(v -> mConsoleImpl.clear());
    }

    public static void launch(@NotNull Context context) {
        context.startActivity(new Intent(context, LogActivity.class)
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_console, menu);
        return true;
    }

    public void export(String fileName) {
        Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        intent.setType("text/plain");
        intent.putExtra(Intent.EXTRA_TITLE, fileName);
        startActivityForResult(intent, RESULT_CODE_EXPORT);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {

        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == RESULT_CODE_EXPORT && resultCode == Activity.RESULT_OK) {
            mConsoleImpl.export(data.getData());
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_copy) {
            mConsoleImpl.copyAll();
        } else if (item.getItemId() == R.id.action_send) {
            mConsoleImpl.send();
        } else if (item.getItemId() == R.id.action_export) {
            mConsoleImpl.export();
        } else if (item.getItemId() == R.id.action_clear) {
            mConsoleImpl.clear();
        } else if (item.getItemId() == R.id.action_settings) {
            // TODO by SuperMonster003 on Jul 3, 2023.
            ViewUtils.showToast(this, R.string.text_under_development_title);
        }
        return true;
    }

}

File: app/src/main/java/org/autojs/autojs/ui/log/LogActivity.kt
package org.autojs.autojs.ui.log

import android.app.ActivityOptions
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import org.autojs.autojs.AutoJs.Companion.instance
import org.autojs.autojs.core.console.GlobalConsole
import org.autojs.autojs.runtime.api.Mime
import org.autojs.autojs.ui.BaseActivity
import org.autojs.autojs.util.ViewUtils
import org.autojs.autojs.util.ViewUtils.showToast
import org.autojs.autojs6.R
import org.autojs.autojs6.databinding.ActivityLogBinding

class LogActivity : BaseActivity() {

    private lateinit var mConsoleImpl: GlobalConsole

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val activityLogBinding = ActivityLogBinding.inflate(layoutInflater)

        setContentView(activityLogBinding.root)
        setToolbarAsBack(R.string.text_log)

        activityLogBinding.console.apply {
            setConsole(instance.globalConsole.also { mConsoleImpl = it })
            setLogActivity(this@LogActivity)
            setPinchToZoomEnabled(true)
        }

        activityLogBinding.fab.apply {
            setOnClickListener { mConsoleImpl.clear() }
            ViewUtils.excludeFloatingActionButtonFromNavigationBar(this)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_console, menu)
        return true
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == RESULT_CODE_EXPORT && resultCode == RESULT_OK) {
            mConsoleImpl.export(data?.data)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_copy -> {
                mConsoleImpl.copyAll()
            }
            R.id.action_send -> {
                mConsoleImpl.send()
            }
            R.id.action_export -> {
                mConsoleImpl.export()
            }
            R.id.action_clear -> {
                mConsoleImpl.clear()
            }
            R.id.action_settings -> {
                // TODO by SuperMonster003 on Jul 3, 2023.
                showToast(this, R.string.text_under_development)
            }
        }
        return true
    }

    @Suppress("DEPRECATION")
    override fun finish() {
        super.finish()
        overridePendingTransition(R.anim.no_anim_fade_in, R.anim.no_anim_fade_out)
    }

    @Suppress("DEPRECATION")
    fun export(fileName: String?) {
        Intent(Intent.ACTION_CREATE_DOCUMENT).apply {
            addCategory(Intent.CATEGORY_OPENABLE)
            setType(Mime.TEXT_PLAIN)
            putExtra(Intent.EXTRA_TITLE, fileName)
            startActivityForResult(this, RESULT_CODE_EXPORT)
        }
    }

    companion object {

        private const val RESULT_CODE_EXPORT = 33128

        @JvmStatic
        fun launch(context: Context) {
            val intent = Intent(context, LogActivity::class.java).apply { addFlags(Intent.FLAG_ACTIVITY_NEW_TASK) }
            val options = ActivityOptions.makeCustomAnimation(context, R.anim.no_anim_fade_in, R.anim.no_anim_fade_out)
            context.startActivity(intent, options.toBundle())
        }

    }

}
